ouroboros-network/src/Ouroboros/Network/KeepAlive.hs

```diff
@@ -8,43 +8,90 @@ module Ouroboros.Network.KeepAlive
   ( KeepAliveInterval (..)
   , keepAliveClient
   , keepAliveServer
+
+  , TraceKeepAliveClient
   ) where
 
-import           Control.Monad.Class.MonadSTM
-import           Control.Monad.Class.MonadTime (DiffTime)
+import           Control.Exception (assert)
+import qualified Control.Monad.Class.MonadSTM as Lazy
+import           Control.Monad.Class.MonadSTM.Strict
+import           Control.Monad.Class.MonadTime
 import           Control.Monad.Class.MonadTimer
+import           Control.Tracer (Tracer, traceWith)
+import qualified Data.Map.Strict as M
 
 import           Ouroboros.Network.Mux (RunOrStop (..), ScheduledStop)
+import           Ouroboros.Network.DeltaQ
 import           Ouroboros.Network.Protocol.KeepAlive.Client
 import           Ouroboros.Network.Protocol.KeepAlive.Server
 
 
 newtype KeepAliveInterval = KeepAliveInterval { keepAliveInterval :: DiffTime }
 
+data TraceKeepAliveClient peer =
+    AddSample peer DiffTime PeerGSV
+
+instance Show peer => Show (TraceKeepAliveClient peer) where
+    show (AddSample peer rtt gsv) = "AddSample " ++ show peer ++ " sample: " ++ show rtt
+        ++ " gsv: " ++ show gsv
 
 keepAliveClient
-    :: forall m.
+    :: forall m peer.
        ( MonadSTM   m
+       , MonadMonotonicTime m
        , MonadTimer m
+       , Ord peer
        )
-    => ScheduledStop m
+    => Tracer m (TraceKeepAliveClient peer)
+    -> ScheduledStop m
+    -> peer
+    -> (StrictTVar m (M.Map peer PeerGSV))
     -> KeepAliveInterval
+    -> StrictTVar m (Maybe Time)
     -> KeepAliveClient m ()
-keepAliveClient shouldStopSTM KeepAliveInterval { keepAliveInterval } =
+keepAliveClient tracer shouldStopSTM peer dqCtx KeepAliveInterval { keepAliveInterval } startTimeV =
     SendMsgKeepAlive go
   where
-    decisionSTM :: TVar m Bool
+    payloadSize = 2
+
+    decisionSTM :: Lazy.TVar m Bool
                 -> STM  m RunOrStop
-    decisionSTM delayVar =
-      do
-       readTVar delayVar >>= fmap (const Run) . check
-      `orElse`
-      shouldStopSTM
+    decisionSTM delayVar = do
+       shouldStop <- shouldStopSTM
+       case shouldStop of
+            Stop -> return Stop
+            Run  -> do
+              done <- Lazy.readTVar delayVar
+              if done
+                 then return Run
+                 else retry
 
     go :: m (KeepAliveClient m ())
     go = do
-      delayVar <- registerDelay keepAliveInterval
+      endTime <- getMonotonicTime
+      startTime_m <- atomically $ readTVar startTimeV
+      case startTime_m of
+           Just startTime -> do
+               let rtt = diffTime endTime startTime
+                   sample = fromSample startTime endTime payloadSize
+               gsv' <- atomically $ do
+                   m <- readTVar dqCtx
+                   assert (peer `M.member` m) $ do
+                     let m' = M.adjust (\a -> a <> sample) peer m
```

Comment:
You could use [updateLookupWithKey](https://hackage.haskell.org/package/containers-0.6.3.1/docs/Data-Map-Strict.html#v:updateLookupWithKey) to both update the dictionary and get the update value.